qemu/tests/numa_opts.py

```diff
@@ -1,49 +1,115 @@
+        MemTotal = get_mem_info(session, 'MemTotal')
+        MemTotal = float(normalize_data_size('%s KB' % MemTotal))
+        error_context.context("MemTotal in guest os is %s MB"
+                              % MemTotal, logging.info)
+        threshold = float(params.get_numeric("mem") - MemTotal) + 30
+        error_context.context("The acceptable threshold is: %s"
+                              % threshold, logging.info)
+    else:
+        numa_guest = numa_monitor
+    session.close()
+
+    monitor_expect_nodes = int(params.get("monitor_expect_nodes", 0))
+    if len(numa_monitor) != monitor_expect_nodes:
+        test.fail("[Monitor]Wrong number of numa nodes: %d. Expected: %d" %
+                  (len(numa_monitor), monitor_expect_nodes))
 
-    for nodenr, node in enumerate(numa):
+    if os_type == 'linux':
+        guest_expect_nodes = int(params.get("guest_expect_nodes",
+                                            monitor_expect_nodes))
+        if len(numa_guest) != guest_expect_nodes:
+            test.fail("[Guest]Wrong number of numa nodes: %d. Expected: %d" %
+                      (len(numa_guest), guest_expect_nodes))
+
+    for nodenr, node in enumerate(numa_guest):
         mdev = params.get("numa_memdev_node%d" % (nodenr))
         if mdev:
             mdev = mdev.split('-')[1]
             size = float(normalize_data_size(params.get("size_%s" % mdev)))
         else:
             size = params.get_numeric("mem")
-        if size != numa[nodenr][0]:
-            test.fail("Wrong size of numa node %d: %d. Expected: %d" %
-                      (nodenr, numa[nodenr][0], size))
+        if len(numa_monitor) != 0 and size != numa_monitor[nodenr][0]:
+            test.fail("[Monitor]Wrong size of numa node %d: %f. Expected: %f" %
+                      (nodenr, numa_monitor[nodenr][0], size))
+        if os_type == 'linux':
+            if size - numa_guest[nodenr][0] > threshold:
+                test.fail("[Guest]Wrong size of numa node %d: %f. Expected: %f"
+                          % (nodenr, numa_guest[nodenr][0], size))
 
         cpus = params.get("numa_cpus_node%d" % (nodenr))
         if cpus is not None:
             cpus = set([int(v) for v in cpus.split(",")])
-            if cpus != numa[nodenr][1]:
+        else:
+            cpus = set([int(v) for v in range(params.get_numeric('smp'))])
+        if len(numa_monitor) != 0 and cpus != numa_monitor[nodenr][1]:
+            test.fail(
+                "[Monitor]Wrong CPU set on numa node %d: %s. Expected: %s" %
+                (nodenr, numa_monitor[nodenr][1], cpus))
+        if os_type == 'linux':
+            if cpus != numa_guest[nodenr][1]:
                 test.fail(
-                    "Wrong CPU set on numa node %d: %s. Expected: %s" %
-                    (nodenr, numa[nodenr][1], cpus))
+                    "[Guest]Wrong CPU set on numa node %d: %s. Expected: %s" %
+                    (nodenr, numa_guest[nodenr][1], cpus))
```

Comment:
Maybe we could adjust the check order so we don't have to check os type so many times.

Reply:
Updated, thanks ~